### PR TITLE
Add tests for checking internal API leakage using ArchUnit

### PR DESCRIPTION
**Pull request by wolfs**
So that we don't add any more leakages.

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by wolfs** [comment]
@bot-gradle test QFL

File: subprojects/architecture-test/build.gradle.kts [hunk 1]
```diff
@@ -29,6 +30,8 @@ tasks.test {
         project.file("src/changes/archunit_store"),
         providers.gradleProperty("archunitRefreeze").map { true })
     )
+
+    useJUnitPlatform()
```
Comment (blindpirate) [inline subprojects/architecture-test/build.gradle.kts:34]: Why is this necessary? Isn't it configured in `gradlebuild.unittest-and-compile.gradle.kts`?

File: subprojects/architecture-test/src/changes/archunit_store/stored.rules [hunk 2]
```diff
@@ -1,4 +1,6 @@
 #
-#Fri Jun 25 10:33:07 CEST 2021
+#Fri Aug 06 17:03:32 CEST 2021
+methods\ that\ are\ public\ API\ methods\ should\ have\ only\ arguments\ or\ return\ types\ that\ are\ Gradle\ public\ API\ or\ primitive\ or\ built-in\ JDK\ classes\ or\ Groovy\ classes=3c3e47be-b5ed-426a-8b9f-cfa00d3b09db
+classes\ that\ are\ Gradle\ public\ API\ should\ not\ have\ direct\ super-class\ or\ interface\ that\ are\ Gradle\ Internal\ API=4b863378-ccfd-4cc1-af5b-e03ff6f3523b
```
Comment (blindpirate) [inline subprojects/architecture-test/src/changes/archunit_store/stored.rules:4]: This is really hard to read... I wonder if there's better ways..

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            return input.isPrimitive();
+        }
+    };
+
+    private static final DescribedPredicate<JavaClass> allowed_types_for_public_api =
+        gradlePublicApi()
+            .or(primitive)
+            .or(resideInAnyPackage("java.lang", "java.util", "java.util.concurrent", "java.util.regex", "java.util.function", "java.lang.reflect", "java.io")
+                .or(type(File[].class))
+                .or(type(FileInputStream.class))
```
Comment (blindpirate) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java:90]: Isn't `type(FileInputStream.class)` duplicate of `resideInAnyPackage("java.io")`?

File: subprojects/architecture-test/build.gradle.kts (hunk 1 above)
Comment (wolfs) [reply subprojects/architecture-test/build.gradle.kts]: You are right, removed!

File: subprojects/architecture-test/src/changes/archunit_store/stored.rules (hunk 2 above)
Comment (blindpirate) [reply subprojects/architecture-test/src/changes/archunit_store/stored.rules:4]: It's fine, I'm just curious.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java (hunk 3 above)
Comment (wolfs) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java]: You are right! I removed the ones already covered by `java.io`.

**Review by blindpirate — APPROVED** [review]
LGTM!

**Comment by wolfs** [comment]
@bot-gradle test and merge
